### ISSUE TITLE
Parallel find_packages_allowing_duplicates?

### ISSUE THREAD

**Issue by mikepurvis** [ros-infrastructure/catkin_pkg#169]
For large workspaces (600+ packages), the serialized parsing of `package.xml` files in `find_packages_allowing_duplicates` can start to be non-trivial:

https://github.com/ros-infrastructure/catkin_pkg/blob/a4cb118d2da9a218f5e5b2216f5b1fdb93c0c5c5/src/catkin_pkg/packages.py#L109-L110

Without changing the interface to the function, would we consider allowing this work to be spread over multiple threads or processes, possibly triggered by some threshold in number of packages?

**Comment by dirk-thomas** [comment]
Absolutely! The caller shouldn't care how the requested information is being gathered. If that loop can be parallelized that would be great.

**Comment by mikepurvis** [comment]
A naive threading implementation is slower than the simple loop, so it's not IO bound. I get ~1.5s with the current implementation, and <0.5s running it with a multiprocessing map. I'll send a PR shortly.

**Comment by mikepurvis** [comment]
The simplest implementation is like so:

```
package_paths = find_package_paths(basepath, exclude_paths=exclude_paths, exclude_subspaces=exclude_subspaces)
processed = multiprocessing.Pool(4).map(parse_package, package_paths)
return dict(zip(package_paths, parsed_packages))
```

However, to preserve the behaviour of the `warnings` attribute requires passing an extra attribute into the map for it and then manually aggregating the results, which necessitates some additional wrapping, unfortunately.

**Comment by dirk-thomas** [comment]
Addressed by #171.
